gnocchi/rest/api.py

```diff
@@ -1515,15 +1515,24 @@ def post(self, create_metrics=False):
         unknown_metrics = []
         unknown_resources = []
         body_by_rid = {}
+
+        attribute_filter = {"or": []}
+        for original_resource_id, resource_id in body:
+            names = list(body[(original_resource_id, resource_id)].keys())
+            attribute_filter["or"].append({"and": [
+                    {"=": {"resource_id": resource_id}},
+                    {"in": {"name": names}}]})
+
+        all_metrics = {}
+        for metric in pecan.request.indexer.list_metrics(
+                attribute_filter=attribute_filter):
+            all_metrics.setdefault(metric.resource_id, []).append(metric)
+
         for original_resource_id, resource_id in body:
             body_by_rid[resource_id] = body[(original_resource_id,
                                              resource_id)]
             names = list(body[(original_resource_id, resource_id)].keys())
-            metrics = pecan.request.indexer.list_metrics(
-                attribute_filter={"and": [
-                    {"=": {"resource_id": resource_id}},
-                    {"in": {"name": names}},
-                ]})
+            metrics = all_metrics[resource_id]
```

Comment:
Agreed with @chungg , this line will fail is no metric have been found for a resource.